[#1710] stop registry to return hashed-passwords details

jbtrystram opened the pull request:
The device registry now only returns metadata and a secret ID for each secret.

Signed-off-by: Jean-Baptiste Trystram <jbtrystram@redhat.com>

sophokles73 commented:
@jbtrystram some integration tests seem to fail due to the changes you have made. Can you check?

jbtrystram commented:
@sophokles73 I am on this today :)

sophokles73 commented on site/documentation/content/api/credentials/index.md:
```diff
@@ -155,6 +155,7 @@ Below is an example for a payload containing [a hashed password]({{< relref "#ha
   "auth-id": "sensor1",
   "enabled": true,
   "secrets": [{
+    "id": "349556ea-4902-47c7-beb0-1009ab693fb4",
```
IMHO there is no need to include the id in the response payload. What should a protocol adapter do with it?

jbtrystram replied:
Indeed, the adapters don't have any use of this. I made a mistake by editing that file ! I reverted the changes

sophokles73 commented on site/documentation/content/api/management/device-registry-v1.yaml:
```diff
@@ -1077,15 +1075,10 @@ components:
+            secrets: [{
+               "id": "349556ea-4902-47c7-beb0-1009ab693fb4",
+               "not-after": "2027-12-24T19:00:00Z",
+               "pwd-plain": ""
```
what is this property good for in this context? Isn't this the response to a *get credentials* request?

jbtrystram replied:
the ID property of the secret is provided here so the user can keep track of the different secrets without handing over the actual values. I also thought the `id` property could be provided in the payload to a `PUT` request on the management API if the user need to update this specific secret. (without creating another one, or submitting all of the other credentials, but that is not implemented yet) .

sophokles73 replied:
I actually meant the empty *pwd-plain* property ...
So, if I get it right, then the *id* property currently has no purpose at all because it is not considered when adding/updating credentials, right?

jbtrystram replied:
Indeed, the `pwd-plain`  does not have any sense here. removed.
the id field still have a use IMHO. When the user set some credentials, he can "verify" they are in the registry with receiving an id for each password. I think if I don't know hono very well I would be confused if I received an empty secrets array after registering a hashed password. With the id field I know the password is there.

sophokles73 commented on site/documentation/content/user-guide/device-registry.md:
```diff
@@ -299,9 +299,9 @@ HTTP/1.1 204 No Content
-This uses a convenient option which lets the Device Registry do the hashing of the password. The following command retrieves the credentials that are stored by the Device Registry as a result of the command above: 
-    
-~~~sh    
+This lets the Device Registry do the hashing of the password. The following command retrieves some metadata credentials that are stored by the Device Registry as a result of the command above: 
```
I would rather keep the original wording and instead explain explicitly, why the response payload does not contain the password itself ...

jbtrystram replied:
I updated the wording. I am not really happy with it though, I am open to suggestions :)

sophokles73 commented on site/homepage/content/release-notes.md:
```diff
@@ -25,6 +25,11 @@ title = "Release Notes"
 
+### API Changes
+
+ * The device registry credentials endpoint will no longer handout details for hashed-password secrets.
```
please be more explicit about which information is no longer returned

jbtrystram replied:
I updated. Let me know what you think.

sophokles73 commented on site/documentation/content/user-guide/device-registry.md:
```diff
@@ -319,6 +319,7 @@ ETag: becc93d7-ab0f-48ec-ad26-debdf339cbf4x
   ]
 }]
 ~~~
+The above result does not contains `pwd_hash` and `hash_function` of the secret, as these value are confidential.
```
... not contain ... as these values ...

sophokles73 commented on site/homepage/content/release-notes.md:
```diff
@@ -27,8 +27,9 @@ title = "Release Notes"
- * The device registry credentials endpoint will no longer handout details for hashed-password secrets.
-   Each password is given an ID which is now returned by the device registry.
+ * The device registry credentials endpoint will no longer handout sensitive details for hashed-password secrets.
+   `pwd-hash`, `Salt` and `hash-function` are stored by the device registry but not returned to the user.
```
`salt`

jbtrystram replied:
Updated, thanks for catching this !

sophokles73 commented on core/src/main/java/org/eclipse/hono/util/RegistryManagementConstants.java:
```diff
@@ -107,6 +107,10 @@
+    /**
+     * The name of the field that contains the secret id.
+     */
+    public static final String FIELD_SECRETS_ID                 = "id";
```
How about renaming this to the more general `FIELD_ID` since it is not really secrets specific anyway and might be used for other entities as well. WDYT?

jbtrystram replied:
Why not. I updated the description so it doesn't lower the code readability.

sophokles73 commented on service-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonSecret.java:
```diff
@@ -55,6 +58,21 @@ public CommonSecret setEnabled(final Boolean enabled) {
+    }
+
+    /**
+     * Sets the ID of the secret. This id may be assigned by the device registry.
```
what is the uniqueness scope of this ID? I assume it is just the containing Credentials instance, right? However, IMHO we should document this here.

jbtrystram replied:
I updated the description.

sophokles73 replied:
Can these IDs be assigned by a client, i.e. when registering the credentials? If so, do we check that the IDs are unique within the credentials?

sophokles73 commented on site/documentation/content/user-guide/device-registry.md:
```diff
@@ -314,23 +314,21 @@ ETag: becc93d7-ab0f-48ec-ad26-debdf339cbf4x
-    
+The above result does not contain `pwd_hash` and `hash_function` of the secret, as these values are confidential.
+
 The following commands add some `hashed-password` credentials for device `4720` using authentication identifier `sensor20`:
```
does this actually *add* credentials or does it *replace* all existing credentials? IMHO it is the latter ...

jbtrystram replied:
At this point of the guide there are no credentials registered for device `4720`, so it actually `add` some. I changed the wording to be more clear though.

sophokles73 commented on site/documentation/content/user-guide/device-registry.md:
```diff
@@ -403,7 +398,7 @@ ETag: d383ba4d-1853-4d03-b322-b7ff5af05ae2
 **Example**
 
-The following command retrieves credentials for device `4720`:
+The following command retrieves credentials metadata for device `4720`:
```
I guess the example payload should at least contain two secrets with e.g. adjoining validity periods so that you can see, what the *id* would be useful for (e.g. for changing one of the two passwords) ...

jbtrystram replied:
Good idea ! I added another secrets with `not-after` and `not-before` overlapping with only one day.

sophokles73 commented on tests/src/test/java/org/eclipse/hono/tests/registry/CredentialsHttpIT.java:
```diff
@@ -403,9 +403,11 @@ public void testGetAddedCredentials(final VertxTestContext context) {
+                    final PasswordCredential cred = b.toJsonArray().getJsonObject(0).mapTo(PasswordCredential.class);
+                    cred.getSecrets().forEach(secret -> {
+                        context.verify(() -> assertThat(secret.getId()).isNotNull());
+                        context.completeNow();;
```
you shouldn't complete the context for each secret ...

jbtrystram replied:
Good catch, thanks

sophokles73 commented on tests/src/test/java/org/eclipse/hono/tests/registry/CredentialsHttpIT.java:
```diff
@@ -541,8 +543,33 @@ public void testGetAddedCredentialsButWithWrongAuthId(final VertxTestContext con
-        expected.stream().forEach(credential -> expectedArray.add(JsonObject.mapFrom(credential)));
+        expected.stream().forEach(credential -> {
+            final JsonObject jsonCredential = JsonObject.mapFrom(credential);
+            expectedArray.add(stripHashAndSaltFromPasswordSecret(jsonCredential));
```
there is not much left to compare, is there?

jbtrystram replied:
the credential type, auth-id and all the credential-level fields are still there

jbtrystram commented:
@sophokles73 for some reason I can't reply to one of your comments in the review.
> I actually meant the empty pwd-plain property ...
> So, if I get it right, then the id property currently has no purpose at all because it is not considered when adding/updating credentials, right?

the `id` field still have a use IMHO. When the user set some credentials, he can "verify" they are in the registry with receiving an `id` for each password. I think if I don't know hono very well I would be confused if I received an empty secrets array after registering a hashed password. With the `id` field I know the password is there.

sophokles73 commented on services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/FileBasedCredentialsService.java:
```diff
@@ -539,6 +539,10 @@ public void set(final String tenantId, final String deviceId, final Optional<Str
             }
             secretsJson.addAll(credentialObject.getJsonArray(CredentialsConstants.FIELD_SECRETS));
+            // add an id for each secret.
+            secretsJson.forEach(secret -> {
```
what if they already have an ID, e.g. provided by the client?

jbtrystram commented:
@sophokles73 @ctron  I implemented the "id" feature for secrets in the file based registry.

It should works as follow : 

- When a user create a secret, the secret is assigned an ID that is returned on `GET` calls.
- A secret can be updated if the corresponding ID is put in the payload on a `PUT` call. A new ID will be assigned.
- If the user provide an Id that doesn't exist it will be ignored.
- On a `PUT` call, all the already existing secrets for which the IDs are not specified in the payload will be deleted.
- - On a `PUT` call, if only the ID is specified in the secret payload, the secret will be kept as-is.

I am going to write a few tests to make sure all works as expected.

ctron commented:
This looks good to me in general. My only comment would be, that I think we should have integration tests for use cases we had in mind. Like 1) Create a new credential set 2) read it back 3) change a single secret 4) verify the update by authenticating a device

Aside from that, but that is not something this PR has touched: I think we should remove the "deprecation" note on the hashed passwords when setting secrets.

jbtrystram commented:
I found a couple of corner cases so I had to rework this . Here are the features now : 

- When a user create a secret, the secret is assigned an ID that is returned on GET calls. A secret can be updated if the corresponding ID is put in the payload on a PUT call. A new ID will be assigned. All the fields that existed before for this ID will be carried over.
- If the user provide an ID that doesn't exist, the secret will be ignored.
- The `PUT` call can also be used to update/add meta-data fields to an existing secret without changing the actual secret values. (e.g. the password). The ID won't be changed as long as the one of the following fields is not changed : `pwd-plain`, `key`, `pwd-hash`, `salt`.
- On a PUT call, if only the ID is specified in the secret payload, the secret will be kept as-is.

It is not possible to delete only one field. The whole secret must be deleted. This can be done with a `PUT` call not containing the ID you want to delete.

jbtrystram commented:
@ctron @sophokles73 I added unit tests to validate the following scenarios : 
- Verify that created secrets contains an ID.
- Verify secrets details are not returned by registry
- Verify one secret can be deleted and the other stays there
- verify that an new ID is assigned when a  secret is changed
- verify the ID doesn't change when metadata only is updated + the other metada that already existed are still there

I think this is ready to be reviewed :) 
I'll squash the commits as soon as it's ready to merge.

sophokles73 commented:
@jbtrystram 

> When a user create a secret, the secret is assigned an ID that is returned on GET calls. A secret can be updated if the corresponding ID is put in the payload on a PUT call. A new ID will be assigned. All the fields that existed before for this ID will be carried over.

I do not quite understand why we would want to assign a new ID with every update. Is there a particular reason for that? The term identifier to suggests that I will be able to refer to the identified object by means of the id for the whole lifespan of the object. So, FMPOV if I change an existing  password, then I would expect the ID to remain the same.

> If the user provide an ID that doesn't exist, the secret will be ignored.

I assume we are still talking about a PUT request. FMPOV it would be more transparent to return a 400 BAD REQUEST in this case in order to not let the client believe that he has actually updated the credentials when in fact we have silently ignored the request ...

jbtrystram commented:
@sophokles73 
 
> I do not quite understand why we would want to assign a new ID with every update. Is there a particular reason for that? The term identifier to suggests that I will be able to refer to the identified object by means of the id for the whole lifespan of the object. So, FMPOV if I change an existing password, then I would expect the ID to remain the same.

My idea was that a new ID would reflect that the change has actually been processed. But I guess that the return code does the same job. I removed this, the ID now stays the same.

> > If the user provide an ID that doesn't exist, the secret will be ignored.
> 
> I assume we are still talking about a PUT request. FMPOV it would be more transparent to return a 400 BAD REQUEST in this case in order to not let the client believe that he has actually updated the credentials when in fact we have silently ignored the request ...

Good point. I updated to return `400 Bad request` and to log the error in that case.

kaniyan commented:
@jbtrystram IMHO also the example credentials (_example-credentials.json_) in the file based device-registry should also be updated in accordance with the changes in this PR. When get credentials operation is invoked for example for device id _4711_ or _gw-1_, it still hands out values such as _pwd-hash_, _hash-function_ etc.

For me it seems that the existing secrets without ids should be migrated by the users so that the credentials endpoint no longer hands out sensitive details. FMPOV it is better to add this info in the release notes too. WDYT?

jbtrystram commented:
> @jbtrystram IMHO also the example credentials (_example-credentials.json_) in the file based device-registry should also be updated in accordance with the changes in this PR. When get credentials operation is invoked for example for device id _4711_ or _gw-1_, it still hands out values such as _pwd-hash_, _hash-function_ etc.

Good catch! That happened only one loading credentials from a file, which i did not tested. Thanks, I fixed it.

> For me it seems that the existing secrets without ids should be migrated by the users so that the credentials endpoint no longer hands out sensitive details. FMPOV it is better to add this info in the release notes too. WDYT?

The endpoint will no longer hand out sensitive details even if the secrets don't contain IDs. That said it would indeed be good to advise users to delete and recreate the credentials entries so the secrets are assigned ids

jbtrystram commented:
@ctron @sophokles73 

Here is what the feature set looks like now after your feedback :
- When a user create a secret, the secret is assigned an ID that is returned on GET calls. A secret can be updated if the corresponding ID is put in the payload on a PUT call. All existing metadata fields will be deleted if not re-supplied in the payload (keeping the get-> put flow).
- If the user provide an ID that doesn't exist, the registry will respond with `400 Bad Request`
- The PUT call can also be used to update/add meta-data fields to an existing secret without changing the actual secret values. (e.g. the password). The following values will be caried over from the existing secret : pwd-plain, key, pwd-hash, salt.

ctron commented:
Sounds right to me. With the exception that the `pwd-plain` should **never** be stored.

jbtrystram commented:
> Sounds right to me. With the exception that the `pwd-plain` should **never** be stored.

It is not ! Sorry my statement was confusing, It's just that this is the values the code check for, as a double check :)

sophokles73 commented:
@jbtrystram  Just for my understanding: I assume that I can add credentials to a set of existing credentials by means of PUTting the payload I got back using  a GET and adding new credentials without an ID to the credentials array, right?

jbtrystram commented:
> @jbtrystram Just for my understanding: I assume that I can add credentials to a set of existing credentials by means of PUTting the payload I got back using a GET and adding new credentials without an ID to the credentials array, right?

of course

dejanb commented on service-base/src/main/java/org/eclipse/hono/service/management/credentials/PasswordSecret.java:
```diff
@@ -112,6 +112,9 @@ protected ToStringHelper toStringHelper() {
     @Override
     public void checkValidity() {
         super.checkValidity();
+        if (containsOnlySecretId()){
```
Can we keep formatting consistent. There are a couple of places where there's no space after bracket and at some places there's a space in negation statement. We can also consider adding it to the check style.

dejanb commented on service-base/src/main/java/org/eclipse/hono/service/management/credentials/PasswordSecret.java:
```diff
@@ -123,6 +126,18 @@ public void checkValidity() {
+     *
+     * @return        true if an ID is present and no new secret detail is set.
+     */
+    public boolean containsOnlySecretId(){
```
Inconsistent formatting.

dejanb commented on services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/FileBasedCredentialsService.java:
```diff
@@ -479,17 +480,19 @@ public void set(final String tenantId, final String deviceId, final Optional<Str
         // authId->credentials[]
         final Map<String, JsonArray> credentialsForTenant = createOrGetCredentialsForTenant(tenantId);
 
+        if (! credentialsForTenant.isEmpty()){
```
Inconsistent formatting.

dejanb commented on services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/FileBasedCredentialsService.java:
```diff
@@ -479,17 +480,19 @@ public void set(final String tenantId, final String deviceId, final Optional<Str
+        if (! credentialsForTenant.isEmpty()){
+            try {
+                verifyOverwriteEnabled(span);
+            } catch (ClientErrorException e){
```
Inconsistent formatting.

dejanb commented on services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/FileBasedCredentialsService.java:
```diff
@@ -804,4 +856,41 @@ public String toString() {
+    /**
+     * Strips the hashed-password details from the jsonObject if needed.
+     */
+    private static void removePasswordDetailsFromCredential(final JsonObject credential){
```
Inconsistent formatting.

dejanb commented on services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/FileBasedCredentialsService.java:
```diff
@@ -804,4 +856,41 @@ public String toString() {
+     * Strips the hashed-password details from the jsonObject if needed.
+     */
+    private static void removePasswordDetailsFromCredential(final JsonObject credential){
+        if (credential.getString(CredentialsConstants.FIELD_TYPE).equals(CredentialsConstants.SECRETS_TYPE_HASHED_PASSWORD)){
```
Inconsistent formatting.

dejanb commented on services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/FileBasedCredentialsService.java:
```diff
@@ -804,4 +856,41 @@ public String toString() {
+        }
+    }
+
+    private static void removePasswordDetailsFromSecret(final JsonObject secret){
```
Inconsistent formatting.

dejanb commented on services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/FileBasedCredentialsService.java:
```diff
@@ -804,4 +856,41 @@ public String toString() {
+        secret.remove(CredentialsConstants.FIELD_SECRETS_KEY);
+    }
+
+    private static void copySecretFields(final JsonObject in, final JsonObject out){
```
Inconsistent formatting.

dejanb commented on services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/FileBasedCredentialsService.java:
```diff
@@ -804,4 +856,41 @@ public String toString() {
+    private void verifyOverwriteEnabled(final Span span) {
+
+        // check if we may overwrite
+        if (! config.isModificationEnabled()) {
```
Inconsistent formatting.

dejanb commented on tests/src/test/java/org/eclipse/hono/tests/registry/CredentialsHttpIT.java:
```diff
@@ -566,4 +593,20 @@ private static PskCredential newPskCredentials(final String authId) {
         return json.getJsonArray(CredentialsConstants.CREDENTIALS_ENDPOINT).getJsonObject(0);
     }
+
+    private static JsonObject stripHashAndSaltFromPasswordSecret(final JsonObject credential){
```
Inconsistent formatting.

dejanb commented on tests/src/test/java/org/eclipse/hono/tests/registry/CredentialsHttpIT.java:
```diff
@@ -541,8 +543,33 @@ public void testGetAddedCredentialsButWithWrongAuthId(final VertxTestContext con
+        });
+
+        // secrets id were added by registry, strip it so we can compare other fields.
+        responseBody.forEach(credential ->{
```
Inconsistent formatting.

dejanb commented on services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/FileBasedCredentialsService.java:
```diff
@@ -529,16 +532,68 @@ public void set(final String tenantId, final String deviceId, final Optional<Str
+                    }
+
+                    // check if the secretID given was found
+                    if (! found) {
```
Inconsistent formatting.

dejanb commented on site/homepage/content/release-notes.md:
```diff
@@ -58,6 +58,12 @@ title = "Release Notes"
 
 ## 1.0.3
 
+### API Changes
```
This should probably go as 1.2.0 release API changes?

jbtrystram commented:
Thanks for your remarks @dejanb . I created a new section for 1.2.0 in the release notes and moved the changes.
